### PR TITLE
Title for push notifications

### PR DESCRIPTION
**Pull request by wkhaksar**
With traccar APIs currently we are sending only "display messages" to Firebase, then it is automatically displayed in receiver's device while the app is in background. 

The conecpt [docs ](https://firebase.google.com/docs/cloud-messaging/concept-options) from Firebase says: FCM automatically displays the message to end-user devices on behalf of the client app.

So, with "display messages" it is good to fullfill the "title" property of the notification object as well, otherwise the notification message will be having only app-icon while displaying the message when the app is in background.

I am thinking, we can define a global title for the notification messages, if it exists then will be transferred along with the message body.

### CODE REVIEW

**Comment by wkhaksar** [comment]
I added "**title**" support for short messages through templates.
Currently it is used with Firebase and Pushover as they are supporting titles, rest other channels are only using "**body**".

File: src/main/java/org/traccar/notification/ShortMessage.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.notification;
+
+public class ShortMessage {
```
Comment (tananaev) [inline src/main/java/org/traccar/notification/ShortMessage.java:19]: Do we need this new model? Now that short and long messages have the same format, I think we should be able to simplify many things.

File: src/main/java/org/traccar/notification/NotificationFormatter.java [hunk 2]
```diff
@@ -63,7 +63,7 @@ public static FullMessage formatFullMessage(long userId, Event event, Position p
         return TextTemplateFormatter.formatFullMessage(velocityContext, event.getType());
     }
 
-    public static String formatShortMessage(long userId, Event event, Position position) {
+    public static ShortMessage formatShortMessage(long userId, Event event, Position position) {
```
Comment (tananaev) [inline src/main/java/org/traccar/notification/NotificationFormatter.java:66]: Same here. We probably don't need a separate method.

File: templates/short/alarm.vm [hunk 3]
```diff
@@ -1 +1,2 @@
+#set($title = "Alarm")
```
Comment (tananaev) [inline templates/short/alarm.vm:1]: Do all these titles match email titles?

File: src/main/java/org/traccar/notification/ShortMessage.java (hunk 1 above)
Comment (wkhaksar) [reply src/main/java/org/traccar/notification/ShortMessage.java]: Fixed.

File: templates/short/alarm.vm (hunk 3 above)
Comment (wkhaksar) [reply templates/short/alarm.vm]: In a new commit, copied subjects from "full" templates into "short" templates.

File: src/main/java/org/traccar/notification/ShortMessage.java (hunk 1 above)
Comment (wkhaksar) [reply src/main/java/org/traccar/notification/ShortMessage.java]: With a new commit, I removed it and renamed the "FullMessage" to a common name "Message" and used it for formatting both full and short messages. Also, with a few other renaming to make them distinguishable.

File: src/main/java/org/traccar/notification/NotificationFormatter.java (hunk 2 above)
Comment (wkhaksar) [reply src/main/java/org/traccar/notification/NotificationFormatter.java]: This has been merged into "FormatMessage" in 9fd24af.

**Comment by wkhaksar** [comment]
In this new commit 9fd24af :
1. Removed the new model "ShortMessage", so now we have only a common model **Message** instead of "FullMessage" and "ShortMessage". 
2. Merged "formatFullMessage" and "formatShortMessage" into a signle method: **formatMessage**.
3. Renamed "Message" models for Firebase and Pushover into **Payload**.
4. Copied subjects from "full templates" into "short templates".

File: src/main/java/org/traccar/notificators/NotificatorTelegram.java [hunk 4]
```diff
@@ -74,15 +75,15 @@ public NotificatorTelegram() {
     private void executeRequest(String url, Object message) {
         Context.getClient().target(url).request()
                 .async().post(Entity.json(message), new InvocationCallback<Object>() {
-            @Override
-            public void completed(Object o) {
-            }
+                    @Override
+                    public void completed(Object o) {
+                    }
```
Comment (tananaev) [inline src/main/java/org/traccar/notificators/NotificatorTelegram.java:80]: Can you please revert these formatting changes.

File: src/main/java/org/traccar/notificators/NotificatorPushover.java [hunk 5]
```diff
@@ -74,23 +77,26 @@ public void sendSync(long userId, Event event, Position position) {
             return;
         }
 
-        Message message = new Message();
```
Comment (tananaev) [inline src/main/java/org/traccar/notificators/NotificatorPushover.java:77]: I think it's better to rename `org.traccar.notification.Message` to something like `NotificationMessage`. That way we don't need to rename everything else.

File: src/main/java/org/traccar/notification/TextTemplateFormatter.java [hunk 6]
```diff
@@ -71,16 +71,12 @@ public static Template getTemplate(String name, String path) {
...
+        String formattedMessage = format(velocityContext, name, templatePath);
+        return new Message((String) velocityContext.get("subject"), formattedMessage);
     }
 
-    public static String formatShortMessage(VelocityContext velocityContext, String name) {
-        return formatMessage(velocityContext, name, "short");
-    }
-
-    private static String formatMessage(
+    private static String format(
```
Comment (tananaev) [inline src/main/java/org/traccar/notification/TextTemplateFormatter.java:79]: Is this still needed as a separate method?

**Review by tananaev — COMMENTED** [review]
I think we are almost there. Added minor comments about naming and formatting.

**Comment by wkhaksar** [comment]
Please have a look from 2c4b4cd, your recommendations has been applied.

File: src/main/java/org/traccar/notification/TextTemplateFormatter.java [hunk 7]
```diff
@@ -71,21 +71,11 @@ public static Template getTemplate(String name, String path) {
...
-
-    private static String formatMessage(
-            VelocityContext velocityContext, String name, String templatePath) {
-
+    public static NotificationMessage formatMessage(VelocityContext velocityContext, String name, String templatePath) {
         StringWriter writer = new StringWriter();
         getTemplate(name, templatePath).merge(velocityContext, writer);
-        return writer.toString();
+        String formattedMessage = writer.toString();
+        return new NotificationMessage((String) velocityContext.get("subject"), formattedMessage);
```
Comment (tananaev) [inline src/main/java/org/traccar/notification/TextTemplateFormatter.java:78]:
```suggestion
        return new NotificationMessage((String) velocityContext.get("subject"), writer.toString());
```

File: src/main/java/org/traccar/notificators/NotificatorFirebase.java [hunk 8]
```diff
@@ -77,15 +83,15 @@ public void sendSync(long userId, Event event, Position position) {
...
-                @Override
-                public void completed(Object o) {
-                }
-
-                @Override
-                public void failed(Throwable throwable) {
-                    LOGGER.warn("Firebase notification error", throwable);
-                }
-            });
+                        @Override
```
Comment (tananaev) [inline src/main/java/org/traccar/notificators/NotificatorFirebase.java:86]: Looks like formatting is not reverted here.

File: src/main/java/org/traccar/notificators/NotificatorFirebase.java [hunk 9]
```diff
@@ -66,8 +69,11 @@ public void sendSync(long userId, Event event, Position position) {
         final User user = Context.getPermissionsManager().getUser(userId);
         if (user.getAttributes().containsKey("notificationTokens")) {
 
+            NotificationMessage shortMessage = NotificationFormatter.formatMessage(userId, event, position,"short");
+
             Notification notification = new Notification();
-            notification.body = NotificationFormatter.formatShortMessage(userId, event, position).trim();
+            notification.title= shortMessage.getSubject();
```
Comment (tananaev) [inline src/main/java/org/traccar/notificators/NotificatorFirebase.java:75]:
```suggestion
            notification.title = shortMessage.getSubject();
```

File: src/main/java/org/traccar/notificators/NotificatorTelegram.java [hunk 10]
```diff
@@ -98,12 +99,14 @@ private LocationMessage createLocationMessage(String messageChatId, Position pos
     @Override
     public void sendSync(long userId, Event event, Position position) {
         User user = Context.getPermissionsManager().getUser(userId);
+        NotificationMessage shortMessage = NotificationFormatter.formatMessage(userId, event, position,"short");
```
Comment (tananaev) [inline src/main/java/org/traccar/notificators/NotificatorTelegram.java:102]:
```suggestion
        NotificationMessage shortMessage = NotificationFormatter.formatMessage(userId, event, position, "short");
```

File: src/main/java/org/traccar/api/resource/PasswordResource.java [hunk 11]
```diff
@@ -53,8 +53,8 @@ public Response reset(@FormParam("email") String email) throws SQLException, Mes
                 Context.getUsersManager().updateItem(user);
                 VelocityContext velocityContext = TextTemplateFormatter.prepareContext(null);
                 velocityContext.put("token", token);
-                FullMessage message = TextTemplateFormatter.formatFullMessage(velocityContext, "passwordReset");
-                Context.getMailManager().sendMessage(userId, message.getSubject(), message.getBody());
+                NotificationMessage fullMessage = TextTemplateFormatter.formatMessage(velocityContext, "passwordReset","full");
```
Comment (tananaev) [inline src/main/java/org/traccar/api/resource/PasswordResource.java:56]:
```suggestion
                NotificationMessage fullMessage = TextTemplateFormatter.formatMessage(velocityContext, "passwordReset", "full");
```

File: src/main/java/org/traccar/notificators/NotificatorFirebase.java (hunk 8 above)
Comment (wkhaksar) [reply src/main/java/org/traccar/notificators/NotificatorFirebase.java:86]: Okat, I will try to merge it from master.

Comment (tananaev) [reply src/main/java/org/traccar/notificators/NotificatorFirebase.java:86]: Why do you need to merge it? Just revert this change.

Comment (wkhaksar) [reply src/main/java/org/traccar/notificators/NotificatorFirebase.java:86]: As it is automatically formatted by IDE so the overall file is affected.

Comment (wkhaksar) [reply src/main/java/org/traccar/notificators/NotificatorFirebase.java]: Theser are reverted in 7d5a18e.

**Comment by wkhaksar** [comment]
So, formatting reverted and your suggested changes has been also accepted.

**Comment by tananaev** [comment]
Looks like checks are failing. Make sure you run `gradle check` locally before pushing the changes.

**Comment by wkhaksar** [comment]
Resolved `gradle check` errors locally in commit c370aa6.

**Comment by tananaev** [comment]
Merged, thank you for the contribution.

**Pull request merged**